modules/karpenter/main.tf

```diff
@@ -1,207 +1,406 @@
 
 locals {
   account_id = data.aws_caller_identity.current.account_id
-  partition  = data.aws_partition.current.partition
   dns_suffix = data.aws_partition.current.dns_suffix
+  partition  = data.aws_partition.current.partition
+  region     = data.aws_region.current.name
 }
 
 ################################################################################
-# IAM Role for Service Account (IRSA)
-# This is used by the Karpenter controller
+# Karpenter controller IAM Role
 ################################################################################
 
 locals {
-  create_irsa      = var.create && var.create_irsa
-  irsa_name        = coalesce(var.irsa_name, "KarpenterIRSA-${var.cluster_name}")
-  irsa_policy_name = coalesce(var.irsa_policy_name, local.irsa_name)
-
+  create_iam_role        = var.create && var.create_iam_role
   irsa_oidc_provider_url = replace(var.irsa_oidc_provider_arn, "/^(.*provider/)/", "")
 }
 
-data "aws_iam_policy_document" "irsa_assume_role" {
-  count = local.create_irsa ? 1 : 0
+data "aws_iam_policy_document" "controller_assume_role" {
+  count = local.create_iam_role ? 1 : 0
 
+  # Pod Identity
   statement {
-    effect  = "Allow"
-    actions = ["sts:AssumeRoleWithWebIdentity"]
+    actions = [
+      "sts:AssumeRole",
+      "sts:TagSession",
+    ]
 
     principals {
-      type        = "Federated"
-      identifiers = [var.irsa_oidc_provider_arn]
+      type        = "Service"
+      identifiers = ["pods.eks.amazonaws.com"]
     }
+  }
 
-    condition {
-      test     = var.irsa_assume_role_condition_test
-      variable = "${local.irsa_oidc_provider_url}:sub"
-      values   = [for sa in var.irsa_namespace_service_accounts : "system:serviceaccount:${sa}"]
-    }
+  # IAM Roles for Service Accounts (IRSA)
+  dynamic "statement" {
+    for_each = var.enable_irsa ? [1] : []
 
-    # https://aws.amazon.com/premiumsupport/knowledge-center/eks-troubleshoot-oidc-and-irsa/?nc1=h_ls
-    condition {
-      test     = var.irsa_assume_role_condition_test
-      variable = "${local.irsa_oidc_provider_url}:aud"
-      values   = ["sts.amazonaws.com"]
+    content {
+      actions = ["sts:AssumeRoleWithWebIdentity"]
+
+      principals {
+        type        = "Federated"
+        identifiers = [var.irsa_oidc_provider_arn]
+      }
+
+      condition {
+        test     = var.irsa_assume_role_condition_test
+        variable = "${local.irsa_oidc_provider_url}:sub"
+        values   = [for sa in var.irsa_namespace_service_accounts : "system:serviceaccount:${sa}"]
+      }
+
+      # https://aws.amazon.com/premiumsupport/knowledge-center/eks-troubleshoot-oidc-and-irsa/?nc1=h_ls
+      condition {
+        test     = var.irsa_assume_role_condition_test
+        variable = "${local.irsa_oidc_provider_url}:aud"
+        values   = ["sts.amazonaws.com"]
+      }
     }
   }
 }
 
-resource "aws_iam_role" "irsa" {
-  count = local.create_irsa ? 1 : 0
+resource "aws_iam_role" "controller" {
```

Comment:
Can we keep previous names for these resources in the submodule? `controller` sounds rather generic and it is not mentioned anywhere in the AWS Premium Support docs AFAIK.

WDYT?

Reply:
this I was going to name `this` (per usual), but the problem I ran into is that I can't have a migration of `this` -> `node`, and `irsa` -> `this` - they would just all converge to conflict on `node` eventually. The module is primarily creating resources for the Karpenter controller, hence the name change to that - and theres one additional role that Karpenter requires which is the node IAM role. So thats where the names came from

And controller is a Kubernetes construct - in the Karpenter docs you'll see the term `Karpenter controller` used quite a bit - 
> As someone using Karpenter, once your Kubernetes cluster and the Karpenter controller are up and running

https://karpenter.sh/docs/

Reply:
Thanks for the explanation. I didn't think about the case that such 2 step renaming can be a problem for Terraform.